'osc login' (actually 'openshift ex login')

fabianofranz opened the pull request:
First pass at the `osc login` command. 

**Status**: this command is available and working as an experimental login support, under `openshift ex login`. Will prompt for username and password (or take from the flags) and print the auth token if it succeeds.

Full login support will involve:

- [x] basic `osc login` command structure
- [x] takes `username` and `password` from flags (or prompt if not provided) and prints out the auth token
- [ ] save token to `.kubeconfig` and assure it's used
- [ ] more flags (e.g. server, cluster, etc), should make use of `.kubeconfig` and save config in it
 - `osc login --context=admin` (makes use of user and cluster defined in the context "admin", set `current-context` and `token`)
 - `osc login --user=myself --server=dev` (makes use of user "personal" and cluster "dev" to log in, set `current-context` and `token`)
- [ ] tests
- [ ] better error handling

Testing:

```
openshift ex login [--username=foo] [--password=bar] [--kubeconfig=./openshift.local.certificates/admin/.kubeconfig]
```
Copy the `.kubeconfig` file to `~/.kube/.kubeconfig` to remove the need of the `--kubeconfig` flag. 

If the server doesn't challenge for a basic auth by default (usually older versions of Origin), make sure you start the server with: 
```
sudo ORIGIN_AUTH_REQUEST_HANDLERS=session,basicauth ORIGIN_GRANT_HANDLER=auto _output/local/go/bin/openshift start --cors-allowed-origins=.*
```

Using [golang.org/x/crypto/ssh/terminal](http://golang.org/x/crypto/ssh/terminal) to password input disabling stty echo, might need to reevaluate this, check platforms support, etc.

fabianofranz commented:
FYI @liggitt @deads2k @smarterclayton

smarterclayton commented:
Can you squash?  Also, why bring in crypto/ssh/terminal?  Does the docker tty package not already provide that?

ncdc commented:
FYI this is how I'm using Docker's terminal code: https://github.com/ncdc/kubernetes/blob/4d409f1f891040ed0759ee1f1c28a4ab86c6832d/pkg/kubectl/cmd/exec.go#L66-L82, if that helps.

ncdc commented:
And here's how Docker disables terminal echo: https://github.com/docker/docker/blob/master/api/client/commands.go#L338-L348

fabianofranz commented:
@ncdc Thanks! Looks like it disables echo on TTY terminals and falls back to regular (password will be shown when typing) on everything else (e.g. Windows), correct? I saw a few similar solutions before taking (golang.org/x/crypto/ssh/terminal)[http://golang.org/x/crypto/ssh/terminal] mostly because of Windows/non-TTY support (although I didn't test on such platforms yet).

ncdc commented:
@fabianofranz the Docker term package supports Windows (https://github.com/docker/docker/blob/master/pkg/term/term_windows.go)

fabianofranz commented:
@ncdc hrm, cool. Let me try that.

ncdc commented:
@fabianofranz I'm planning on using Docker's term package with my remote command execution PR to Kubernetes upstream (adding Docker's term package to Godeps). If we find out that doesn't suit your needs for some reason, we should find a package that works for both of us.

fabianofranz commented:
@ncdc I removed golang.org/x/crypto/ssh/terminal and changed it to use github.com/docker/docker/pkg/term. Seems to work pretty well, so let's go with it and I'll take some time to check how it goes on Windows. Thanks.

fabianofranz commented:
@smarterclayton Squashed and removed crypto/ssh/terminal in favor of Docker term.

smarterclayton commented:
@deads2k please review this for consistency with the upstream stuff

jwforres commented:
@fabianofranz so the command doesn't let you specify --hostname or --server, is that coming from kubeconfig?

deads2k commented on pkg/cmd/experimental/tokens/whoami.go:
```diff
@@ -29,14 +37,13 @@ func NewCmdWhoAmI(clientCfg *clientcmd.Config) *cobra.Command {
-func whoami(token string, clientCfg *clientcmd.Config) {
-	osCfg := clientCfg.OpenShiftConfig()
+func whoami(token string, clientCfg *kclient.Config, cmd *cobra.Command) {
 	// This will be pulled out of the auth config file after https://github.com/GoogleCloudPlatform/kubernetes/pull/2437
```
Go ahead and do this cleanup while you're here.  The token from the bound clientConfig should be used.

fabianofranz replied:
Not bound automatically in a first try so I'm adding a TODO to investigate later.

----- Original Message -----
> > @@ -29,14 +37,13 @@ func NewCmdWhoAmI(clientCfg *clientcmd.Config)
> > *cobra.Command {
> >  	return cmd
> >  }
> >  
> > -func whoami(token string, clientCfg *clientcmd.Config) {
> > -	osCfg := clientCfg.OpenShiftConfig()
> > +func whoami(token string, clientCfg *kclient.Config, cmd *cobra.Command) {
> >  	// This will be pulled out of the auth config file after
> >  	https://github.com/GoogleCloudPlatform/kubernetes/pull/2437
> 
> Go ahead and do this cleanup while you're here.  The token from the bound
> clientConfig should be used.
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/830/files#r24107268

fabianofranz commented:
@jwforres It does support `--server` from the global flags, and it can also come from `kubeconfig`. But yeah, the next steps will involve a better integration with configs and maybe make these flags local.

deads2k commented on pkg/cmd/experimental/login/login.go:
```diff
@@ -0,0 +1,62 @@
+	return cmds
+}
+
+// Copy of kubectl/cmd/DefaultClientConfig, using NewNonInteractiveDeferredLoadingClientConfig
```
It feels like we have this somewhere else in openshift as well.  Can you make a TODO to find and merge the duplicates?

fabianofranz replied:
Correct, I found 3 places at least. TODO added.

----- Original Message -----
> > +
> > +			accessToken, err := tokencmd.RequestToken(clientCfg, os.Stdin,
> > usernameFlag, passwordFlag)
> > +			if err != nil {
> > +				glog.Fatal("%v\n", err)
> > +			}
> > +
> > +			fmt.Printf("Auth token: %v\n", string(accessToken))
> > +		},
> > +	}
> > +
> > +	cmds.Flags().StringP("username", "u", "", "Username, will prompt if not
> > provided")
> > +	cmds.Flags().StringP("password", "p", "", "Password, will prompt if not
> > provided")
> > +	return cmds
> > +}
> > +
> > +// Copy of kubectl/cmd/DefaultClientConfig, using
> > NewNonInteractiveDeferredLoadingClientConfig
> 
> It feels like we have this somewhere else in openshift as well.  Can you make
> a TODO to find and merge the duplicates?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/830/files#r24107701

deads2k commented:
Looks like a start.  I'm looking to the next two bullet points.  I'd also like to see login handle the case where you specify a --server, successfully login, look for a matching --cluster (create one if needed), create a new --user, create  matching --context, and set the current-context.  No pressure.  :)

fabianofranz commented:
@smarterclayton @deads2k comments addressed.

smarterclayton commented:
```
$ openshift ex login
[Feb 15 15:24 ~/projects/go/src/github.com/openshift/origin]$ openshift ex login
F0204 15:24:16.494395   88531 login.go:40] %v
Unable to get token.  Try visiting https://localhost:8443/oauth/token/request for a new token.
```

What is that %v?

fabianofranz commented:
oops, fixed

----- Original Message -----
> ```
> $ openshift ex login
> [Feb 15 15:24 ~/projects/go/src/github.com/openshift/origin]$ openshift ex
> login
> F0204 15:24:16.494395   88531 login.go:40] %v
> Unable to get token.  Try visiting https://localhost:8443/oauth/token/request
> for a new token.
> ```
> 
> What is that %v?
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/830#issuecomment-72931652

smarterclayton commented:
Also, the fact that this fails means whatever you're doing for prompting doesn't work on OSX

smarterclayton commented:
IF the server doesn't respond I get the generic error.  I need to get a real message.

jwforres commented:
@fabianofranz I get an error when trying this https://gist.github.com/jwforres/2d4950108ac829bde3bc
According to @liggitt you are probably following the redirect to /oauth/token/display#access_token=....

jwforres commented:
So is login actually storing the token onto the current kubeconfig context (which is what we want)?  Or is it only printing out a token for you to manually pass in on the --token argument?

fabianofranz commented:
@jwforres Ah, that was my last change to make it more verbose on server errors. Let me fix that.
That's right, only printing now. Next step will be to make it save the token (and user, certs, etc if needed) to `.kubeconfig`.

smarterclayton commented:
How hard is it to save the token right now?


> On Feb 4, 2015, at 6:00 PM, Fabiano Franz <notifications@github.com> wrote:
> 
> @jwforres Ah, that was my last change to make it more verbose on server errors. Let me fix that.
> That's right, only printing now. Next step will be to make it save the token (and user, certs, etc if needed) to .kubeconfig.
> 
> —
> Reply to this email directly or view it on GitHub.
>

fabianofranz commented:
@jwforres Fixed.

fabianofranz commented:
I don't hink it's simple like in "done for 1 hour ago". :)
@deads2k thoughts?

----- Original Message -----
> How hard is it to save the token right now?
> 
> 
> > On Feb 4, 2015, at 6:00 PM, Fabiano Franz <notifications@github.com> wrote:
> > 
> > @jwforres Ah, that was my last change to make it more verbose on server
> > errors. Let me fix that.
> > That's right, only printing now. Next step will be to make it save the
> > token (and user, certs, etc if needed) to .kubeconfig.
> > 
> > —
> > Reply to this email directly or view it on GitHub.
> > 
> 
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openshift/origin/pull/830#issuecomment-72959520

deads2k commented:
> How hard is it to save the token right now?

If we limit scope to things that will probably work for most users right now for the first attempt:
 1.  Only modify the local .kubeconfig file.
 1.  Always create a new identity to hold the token with a deconflicted name
 1.  Always create a new context to express the new identity cluster combination
 1.  Always create a new cluster if no --cluster was specified or always reference the existing cluster if it is specified.
 1.  Always update current-context to point to the new context.

I'd put the work at a couple hours.  Keep in mind that's not the nicest user experience, just one that will work in simple cases.

deads2k commented:
Opened a pull request to your branch with a very quick and dirty way of modifying .kubeconfig.  I enumerated it's flaws in the description:  https://github.com/fabianofranz/origin/pull/31

fabianofranz commented:
@deads2k Thanks for the PR, it's a nice start for at least having a glance about login saving to `.kubeconfig`.

fabianofranz commented:
@smarterclayton @deads2k I'd like to get this merged, any other comments?

smarterclayton commented:
Rebase to get rid of the merge commit and I'll merge you.

fabianofranz commented:
@smarterclayton done

smarterclayton commented:
[merge]
